File: src/test/java/com/raychenon/leetcode/tree/TreeUtil.java
package com.raychenon.leetcode.tree;

import java.util.List;

/**
 * User: raychenon
 * Date: 27/6/2020
 * Time: 5:10 PM
 */
public class TreeUtil {

    /**
     * Input: [4,9,0,5,1]
     * --- 4
     * ---/ \
     * --9   0
     * -/ \
     * 5   1
     *
     * @param values
     * @return
     */
    public static TreeNode createTree(List<Integer> values) {
        if (values == null || values.size() == 0) return null;
        TreeNode root = createTree(values, 0);
        return root;
    }

    private static TreeNode createTree(List<Integer> values, int index) {
        if (index >= values.size()) return null;

        Integer value = values.get(index);
        if (value == null) return null;

        TreeNode tree = new TreeNode(value);

        // tree(index).left = 2 * index + 1
        tree.left = createTree(values, index * 2 + 1);

        // tree(index).right = 2 * index + 2
        tree.right = createTree(values, index * 2 + 2);

        return tree;
    }

}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/tree/TreeUtil.kt
package com.raychenon.kotlin.leetcode.tree

import com.raychenon.leetcode.tree.TreeNode

/**
 * User: raychenon
 * Date: 5/7/2020
 * Time: 10:49 PM
 */
object TreeUtil {

    /**
     * Input: [4,9,0,5,1]
     * --- 4
     * ---/ \
     * --9   0
     * -/ \
     * 5   1
     *
     * @param values
     * @return
     */
    fun createTree(vararg nums: Int?): TreeNode? {
        val values = asList(*nums)
        return if (values == null || values.size == 0) null else createTree(values, 0)
    }

    /**
     * similar to kotlin.collections.listOf
     *
     * help from https://stackoverflow.com/questions/44090019/how-to-pass-vararg-as-array-to-function-in-kotlin
     */
    private fun <Int> asList(vararg ts: Int?): List<Int?> {
        val result = mutableListOf<Int?>()
        for (t in ts) result.add(t)
        return result
    }

    private fun createTree(values: List<Int?>, index: Int): TreeNode? {
        if (index >= values.size) return null
        val value = values[index] ?: return null
        val tree = TreeNode(value)

        // tree(index).left = 2 * index + 1
        tree.left = createTree(values, index * 2 + 1)

        // tree(index).right = 2 * index + 2
        tree.right = createTree(values, index * 2 + 2)
        return tree
    }
}